src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,140 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int INIT_SIZE = 10;
+    private static final double ZOOM_SIZE = 1.5;
+    private static final String BUG_INDEX = "The specified index does not exist";
+    private static final String BUG_FOUND = "Specified element not found";
+    private Object[] elementData = new Object[INIT_SIZE];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        if (elementData.length == size) {
+            resize(lengthZoom());
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        if (elementData.length == size) {
+            resize(lengthZoom());
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int numNew = list.size();
+        if (numNew == 0) {
+            return;
+        }
+        if (numNew > (elementData.length - size)) {
+            resize((size + numNew));
+        }
+        for (int i = 0; i < numNew; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        final Object[] es = elementData;
+        T oldValue = (T) es[index];
+        fastRemove(es, index);
```

Comment:
```suggestion
        T oldValue = (T) elementData[index];
        fastRemove(es, index);
```
No need to put the logic in fastRemove write it here